flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -149,16 +154,58 @@ void assignExclusiveSegments() throws IOException {
 		}
 	}
 
+	/**
+	 * Reads the channel state data executed by netty thread, so it can make use of almost all the
+	 * existing processes to avoid bringing additional race conditions with task thread. Also it can
+	 * avoid introducing another thread pool to handle this work to make things more complex.
+	 */
+	private void readInputChannelState() throws IOException {
+		while (true) {
+			Buffer buffer;
+			synchronized (bufferQueue) {
+				buffer = bufferQueue.takeBuffer();
+				if (buffer == null) {
+					if (isReleased()) {
+						return;
+					}
+
+					buffer = inputGate.getBufferPool().requestBuffer();
+					if (buffer != null) {
+						bufferQueue.addFloatingBuffer(buffer);
+						continue;
+					} else {
+						inputGate.getBufferProvider().addBufferListener(this);
+						isWaitingForStateBuffers = true;
+						return;
+					}
+				}
+			}
+
+			ChannelStateReader.ReadResult result = inputGate.stateReader.readInputData(channelInfo, buffer);
```

Comment:
`stateReader.readInputData` is not thread safe as it is currently and this `readInputChannelState` method can be executed by netty threads concurrently for different channels, right?

Reply:
Yes, `readInputChannelState` can be executed concurrently by multiple netty threads for different channels by design. I think in general task processing should be more faster than reading states, so one thread might not be enough for filling buffer to feed task thread well. And every channel actually has exclusive buffers which can be used in parallel to speed up recovery process.

I overlooked the `NotThreadSafe` annotation in `ChanelStateReaderImpl`.  Since every input channel handle will actually generate a separate `ChannelStateStreamReader` and respective stream, I was supposed one input channel state should not be read by multiple threads, but different channel states can be read by different threads concurrent. I would further confirm with Roman whether there are other limitations.